### PR TITLE
Add inference helpers & tests

### PR DESCRIPTION
**Pull request by palp**
This change adds basic inference tests to cover the supported models & samplers, based on the streamlit demo. In service of this, a helper module was extracted from streamlit_helpers.py using kwargs in place of UI elements. A future PR could refactor the streamlit code to use it, avoiding the duplication of code and solidifying a source of truth for inference functions.

### CODE REVIEW

**Comment by akx** [comment]
I also have a minimal txt2img example in a branch in my repo, but it would require some of my other PRs to get merged. Could someone from Stability maybe take a look at them?

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:88]: `msg` is never assigned any other value than `None` – it's a bit weird that this returns it in the first place.

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:102]: See #49 – it's probably not a good idea to hard-code CUDA. (Repeated for other `"cuda"` references.)

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:107]:
```suggestion
    return list({x.input_key for x in conditioner.embedders})
```

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:179]: Should these use those commented values? I think the rounded values are for the streamlit UI.

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:436]: This could be replaced with a single dict construction – `value_dict` and `init_dict` will be the same object anyway.

File: .github/workflows/test-inference.yml
Comment (akx) [inline .github/workflows/test-inference.yml:12]:
```suggestion
    # This action is designed only to run on the Stability research cluster at this time, so many assumptions are made about the environment

```

File: pyproject.toml
Comment (akx) [inline pyproject.toml:47]: `torchaudio` isn't required by the repo as far as I'm aware.

File: .github/workflows/test-inference.yml
Comment (akx) [inline .github/workflows/test-inference.yml:22]: Might want to cache the dependencies (though this might be unnecessary if your runner already has a good local cache)
```suggestion
          python-version: "3.10"
          cache: pip
          cache-dependency-path: requirements-pt2.txt
```

File: pyproject.toml
Comment (akx) [inline pyproject.toml:48]: Most of these requirements really aren't needed if #44 and #50 get merged – I have identified a minimal requirements set, but that's not much use until those are merged.

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:122]: This class doesn't seem to be used for anything?

File: .github/workflows/test-inference.yml
Comment (benjaminaubin) [inline .github/workflows/test-inference.yml:18]: @palp do you want to commit this path?

File: pyproject.toml
Comment (benjaminaubin) [reply pyproject.toml:47]: true this not used

File: pyproject.toml
Comment (benjaminaubin) [reply pyproject.toml:48]: The current PR has to be merged urgently. But we do not forget your contributions that we should review and merge. Thank you for all of this

File: sgm/inference/helpers.py
Comment (benjaminaubin) [inline sgm/inference/helpers.py:1]: if you add these helpers to sgm, we should remove them from https://github.com/Stability-AI/generative-models/blob/main/scripts/demo/streamlit_helpers.py

and adapt https://github.com/Stability-AI/generative-models/blob/main/scripts/demo/sampling.py

otherwise this looks duplicated

Comment (benjaminaubin) [reply sgm/inference/helpers.py:1]: this file can be cleaned but not the goal of the PR

File: sgm/inference/helpers.py
Comment (benjaminaubin) [reply sgm/inference/helpers.py:88]: we shall clean the file in another PR

**Review by benjaminaubin — CHANGES_REQUESTED** [review]
@palp please make sure there is no duplication between 
- scripts/demo
- sgm/inference
- test/inference

Otherwise the other changes will be accepted.
Improvements and cleaning from other PRs should be worked out asap (thank you @akx)

File: sgm/inference/helpers.py
Comment (akx) [reply sgm/inference/helpers.py:1]: It will maybe take quite some additional effort to clean up the duplication cf. streamlit helpers, since those versions use streamlit's widget functions in conditional blocks, e.g.

https://github.com/Stability-AI/generative-models/blob/e25e4c0df1d01fb9720f62c73b4feab2e4003e3f/scripts/demo/streamlit_helpers.py#L249-L252

only uses/shows that widget when that selection is done, and expressing that in a way that lets one use these inference helpers with streamlit doesn't quite sound straightforward.

If you're going to defer cleaning this file up, it's probably not worth to look at that duplication right now either :)

File: sgm/inference/helpers.py
Comment (akx) [reply sgm/inference/helpers.py:88]: TBH, if you're concerned about duplication, the canonical version (the one I fixed in #43) of this function is already in `sgm/util.py`.

File: pyproject.toml
Comment (palp) [reply pyproject.toml:47]: It's present in the requirements.txt, which is why it's here - could be removed from both, but dependency cleanup might be best left for another PR.

File: .github/workflows/test-inference.yml
Comment (palp) [reply .github/workflows/test-inference.yml:22]: Yeah, these runners are not that ephemeral so it has a good chance of being on disk. We also have our own pass-through cache I can hook up that will keep it within AWS.

File: .github/workflows/test-inference.yml
Comment (palp) [reply .github/workflows/test-inference.yml:18]: Yes, we'll revise it soon as we'll set up a dedicated user/account for the runners.

File: sgm/inference/helpers.py
Comment (palp) [reply sgm/inference/helpers.py:1]: To @akx's point it is more complicated, but that may mean the helpers need to change in some meaningful way and it might be better to deal with it now to avoid breaking changes once these are live, since the idea is for people to use them. I will take a crack at it this morning, if it gets too hairy we can still wait.

File: pyproject.toml
Comment (palp) [reply pyproject.toml:48]: Indeed, part of the motivation for this PR is to make it easier to test contributions like yours.

File: .github/workflows/test-inference.yml
Comment (palp) [reply .github/workflows/test-inference.yml:22]: Actually, I've changed my mind and am moving this to a shared path now, which we'll point to via a secret - also making the action a bit closer to being reusable elsewhere.

File: .github/workflows/test-inference.yml
Comment (palp) [reply .github/workflows/test-inference.yml:18]: Actually, I've changed my mind and am moving this to a shared path now, which we'll point to via a secret - also making the action a bit closer to being reusable elsewhere.

File: pyproject.toml
Comment (akx) [reply pyproject.toml:47]: Sure. Dependency cleanup is pending on #44 and #50.

File: sgm/inference/helpers.py
Comment (akx) [reply sgm/inference/helpers.py:1]: For a repo whose version is 0.0.1 I don't think some possible future breaking changes in helpers are that bad 😁

File: sgm/inference/helpers.py
Comment (palp) [reply sgm/inference/helpers.py:122]: > This class doesn't seem to be used for anything?

It is present for downstream usage (same as `apply_refiner`), I will be adding tests for both.

File: sgm/inference/helpers.py
Comment (palp) [reply sgm/inference/helpers.py:1]: > if you add these helpers to sgm, we should remove them from https://github.com/Stability-AI/generative-models/blob/main/scripts/demo/streamlit_helpers.py
> 
> and adapt https://github.com/Stability-AI/generative-models/blob/main/scripts/demo/sampling.py
> 
> otherwise this looks duplicated

This is for the most part done. There are new enums and classes in `api.py` that are not used by streamlit (yet), which avoids rewriting the UI code. But almost all inference code is now shared between `api.py` and `sampling.py` through the `helpers.py` module.

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:54]: These could maybe be just vanilla `@dataclass`es so you wouldn't need the Pydantic dependency.

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:180]: All devices might not support `half` (I've had trouble with that on `mps`), but that might be a future improvement...

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:320]:
```suggestion
        raise ValueError(f"unknown discretization {params.discretization}")

```

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:280]: These `get_*_config` functions would be simplified if each branch just `return`ed the config dict they return instead of assigning to a local and returning at the end.

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:328]:
```suggestion
    if params.sampler in (Sampler.EULER_EDM,Sampler.HEUN_EDM):
```
That said, this is unnecessary since there's a nested `if` for the two anyway.

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:354]:
```suggestion
    elif params.sampler in (Sampler.EULER_ANCESTRAL, Sampler.DPMPP2S_ANCESTRAL):
```
But it's also unnecessary like the above wrapping `if`.

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:324]: This too could be simplified if each branch returned the `sampler` itself.

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:121]: Might be worth it to inline this variable alias – `with autocast` immediately tells seasoned torchistas what's happening, `with precision_scope` might not. (Repeated for `do_img2img`.)

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:137]: In line with #42, these (and other `print`s) should probably be `logger.debug()`s or similar (if they should be here at all).

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:261]: This entire function is already annotated with  `torch.no_grad`, so this (or the annotation) seems superfluous.

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:309]: I don't think `filter` really belongs in `do_img2img` – you can call a function on your samples afterwards, and what if you wanted to filter the latents too?

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:257]: Maybe the `logger` shouldn't be passed in here, but it should just be a module-level constant like #42 does.

That would make it easy to `logger.getLogger(...).setLevel` it out in userland code. (Not printing the sigmas to the Streamlit demo UI doesn't sound like a great loss.)

File: tests/inference/test_inference.py
Comment (akx) [inline tests/inference/test_inference.py:25]: Nifty pytesting here 😁 ✨

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:171]: BTW, is there ever a good reason to _not_ return the latents? It would simplify the signature of this function. (That said, there's currently no way to return latents only and not decode them...)

File: sgm/inference/helpers.py
Comment (akx) [inline sgm/inference/helpers.py:174]: `N` can never be a `ListConfig`, btw – from my experience with #60, `get_batch` is always called in a way that `num_samples` is a list of 1 integer.

File: sgm/inference/api.py
Comment (akx) [inline sgm/inference/api.py:196]: Could be replaced with `value_dict={**dict(params), "prompt": ..., ...}` (repeated for other constructs like this).

**Review by akx — COMMENTED** [review]
Some comments (but if this API's signatures can change in the future, some of them aren't interesting).

I'd maybe recommend also making top-level apis kwarg-only with (e.g.) `do_img2img(*, img, ...)` so user code can't mess up with e.g. parameter order.

File: sgm/inference/api.py
Comment (palp) [reply sgm/inference/api.py:54]: I think I'm on board with this just because having things named `Model` is confusing here.

File: sgm/inference/helpers.py
Comment (palp) [reply sgm/inference/helpers.py:257]: Agreed, it's not worth it for the one line, I'll just remove it for now. Thought there would be more when I put it in.

**Review by benjaminaubin — APPROVED** [review]
Looks good to me

**Pull request merged**
